### PR TITLE
[Product Import] Rename enterprise fields and update template

### LINKED ISSUE #2481: [Product Import] Review csv template on landing page
## Description
We have review the UX of product import to avoid confusion, reduce the number of clics, and try to find some consistency.
Here we want to tackle the template part which was before included in issue #2455
The goal is to have csv template with all possible columns.

## Acceptance criteria

Templates for product listings and inventories are accurate. Especially, the template for inventory only has the necessary columns, some existing columns can be delete. Some column headers have been renamed for more clarity. See https://docs.google.com/document/d/1QDtKQe64enO5ZLnht9PxQ8bOlaMHCueJBMWSyW4zCng/edit# for headers in each template.

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

Closes #2481

As part of #2481 it was decided to rename the names of the fields used in the CSV import. For importing to product list, the enterprise field has changed form "supplier" to "producer", and for inventory the fields "supplier" and "producer" are now "producer" and "distributor".

I've tried to make some of the naming more generic in the codebase, using "enterprise" where possible for lists such as "enterprise_ids" which will now be used either to check the "producer" for product list, or "distributor" for inventory.

#### What should we test?

Product Import generally, using the new names, and the downloadabe CSV templates.

#### Release notes

Renamed CSV fields in product import for clarity and updated CSV template.

Changelog Category: Changed

#### Documentation updates

Product Import guide will need to be updated to reflect the new field names.

### CODE REVIEW

**Comment by luisramos0** [comment]
Lots of "line is too long" errors on codeclimate. Are we really enforcing 80 chars per line? I think 80 is too short. I prefer 100.

**Comment by Matt-Yorkley** [comment]
Some of these lines are starting to look really ugly with the 80 character limit...

File: app/models/product_import/entry_validator.rb [hunk 1]
```diff
@@ -74,32 +74,48 @@ def enterprise_validation(entry)
       return if permissions_error entry
       return if primary_producer_error entry
 
-      entry.enterprise_id = @spreadsheet_data.enterprises_index[entry.enterprise][:id]
-      entry.public_send "#{enterprise_field}_id=", @spreadsheet_data.enterprises_index[entry.enterprise][:id]
+      entry.enterprise_id = \
```
Comment (luisramos0) [inline app/models/product_import/entry_validator.rb]: is this slash needed?

Comment (Matt-Yorkley) [reply app/models/product_import/entry_validator.rb]: Definitely not!

File: app/models/product_import/entry_processor.rb [hunk 2]
```diff
@@ -135,7 +135,7 @@ def save_new_inventory_item(entry)
         @inventory_created += 1
         @updated_ids.push new_item.id
       else
-        @importer.errors.add("#{I18n.t('admin.product_import.model.line')} #{line_number}:", new_item.errors.full_messages)
+        @importer.errors.add("#{I18n.t('admin.product_import.model.line')} #{entry.line_number}:", new_item.errors.full_messages)
```
Comment (mkllnk) [inline app/models/product_import/entry_processor.rb]: When duplicating the same code three or more times, I tend to extract it.

I would also use I18n for the string interpolation. Some languages may prefer something like "nth line:" instead of "Line n:".
```diff
-# with en.yml:
-#   line: Line
-"#{I18n.t('admin.product_import.model.line')} #{entry.line_number}:"

+# with en.yml:
+#   line: "Line %{number}:"
+I18n.t('admin.product_import.model.line', number: entry.line_number)
```

File: app/models/product_import/entry_processor.rb [hunk 3]
```diff
@@ -166,6 +166,7 @@ def save_new_product(entry)
 
       product = Spree::Product.new
       product.assign_attributes(entry.attributes.except('id'))
+      product.supplier_id = entry.enterprise_id
```
Comment (mkllnk) [inline app/models/product_import/entry_processor.rb]: The commit message doesn't explain why this change is needed. Is it necessary because of renaming supplier to enterprise? Or is it fixing a bug?

File: spec/features/admin/product_import_spec.rb [hunk 4]
```diff
@@ -33,7 +33,7 @@
 
     it "validates entries and saves them if they are all valid and allows viewing new items in Bulk Products" do
       csv_data = CSV.generate do |csv|
-        csv << ["name", "supplier", "category", "on_hand", "price", "units", "unit_type"]
+        csv << ["name", "enterprise", "category", "on_hand", "price", "units", "unit_type"]
```
Comment (mkllnk) [inline spec/features/admin/product_import_spec.rb]: I think it's good practice to have code and spec changes in the same commit. It makes the use of git bisect possible and shows the impact of a code change.

File: app/models/product_import/spreadsheet_data.rb [hunk 5]
```diff
@@ -33,10 +34,14 @@ def shipping_index
 
     private
 
+    def import_into_inventory?
```
Comment (mkllnk) [inline app/models/product_import/spreadsheet_data.rb]: We have some code duplication here. I'm wondering if the `entry` could know if it holds a producer or distributor.

File: app/models/product_import/entry_processor.rb [hunk 6]
```diff
@@ -46,7 +46,10 @@ def count_existing_items
 
         products_count =
           if settings.importing_into_inventory?
-            VariantOverride.where('variant_overrides.hub_id IN (?)', enterprise_id).count
+            VariantOverride.where(
```
Comment (mkllnk) [inline app/models/product_import/entry_processor.rb]: I thought `VariantOverride.where(hub_id: [enterprise_id])` would work as well. I just checked the class and this implementation is used for a scope already. You can write `VariantOverride.for_hubs([enterprise_id])`.

File: app/models/product_import/spreadsheet_data.rb (hunk 5 above)
Comment (Matt-Yorkley) [reply app/models/product_import/spreadsheet_data.rb]: I've removed this method, as we already have it in the settings object. I'll call it from there instead, as it's now done that way in most places where that check is made.

**Comment by mkllnk** [comment]
Great. What are your thoughts on my I18n comment?

**Comment by sauloperez** [comment]
> Some of these lines are starting to look really ugly with the 80 character limit...

That was the default but we can increase it. What about 89? I believe that is the ancient standard-ish (my Vim highlights the 89th column).

**Comment by Matt-Yorkley** [comment]
> I believe that is the ancient standard

@sauloperez What's the significance of these ancient standards? Did Moses come down the mountain bearing stone tablets with strictly 89 characters per line? :trollface:

File: app/models/product_import/entry_processor.rb (hunk 2 above)
Comment (Matt-Yorkley) [reply app/models/product_import/entry_processor.rb]: Yep, good thinking.

File: app/models/product_import/entry_processor.rb (hunk 3 above)
Comment (Matt-Yorkley) [reply app/models/product_import/entry_processor.rb]: It's because of the renaming, yes. The product object here needs the `supplier_id` to be set.

File: app/models/product_import/entry_processor.rb (hunk 2 above)
Comment (Matt-Yorkley) [reply app/models/product_import/entry_processor.rb]: I realised I've addressed this here and elsewhere in another PR. I'll remove this commit from here and address it (including your I18n suggestion) in https://github.com/openfoodfoundation/openfoodnetwork/pull/2842

**Comment by sauloperez** [comment]
Well, you're right @Matt-Yorkley :trollface: It's the fear I share with @mkllnk of ending up with too much logic in a line (and that doesn't fit my small laptop screen).

**Comment by sauloperez** [comment]
Staged on https://staging.katuma.org/

**Comment by RachL** [comment]
@Matt-Yorkley I have troubles with shipping_category. In user guide it says that the column is not required, but I can't upload of file without it anymore. Plus, if I had it but leave blank as my 2 products have shipping category set to "none", the error stays the same....
Testing notes:
https://docs.google.com/document/d/1QYAXEXDJE8mlGLSNl30t9W7chhh_X6WQMt4tUgHMF4o/edit#

**Comment by sauloperez** [comment]
Sorry guys I broke Katuma's staging while trying things out for `ofn-install`. I'll see if I can fix it soon. Meanwhile



Sorry for the inconvenience.

**Comment by Matt-Yorkley** [comment]
@RachL I've re-tested this and these are my notes from 4 cases:

- Updating an existing product with no shipping_category column in CSV and shipping_category set to 'None' on the existing product: Success
- Updating an existing product with no shipping_category column in CSV and shipping_category set to 'Frozen' on the existing product: Error
- Updating existing product with empty shipping_category column in CSV and shipping_category set to 'None' on the existing product: Success
- Updating existing product with empty shipping_category column in CSV and shipping_category set to 'Frozen' on the existing product: Error

I think that's correct, no?

**Comment by RachL** [comment]
@Matt-Yorkley Yes this is what I was expecting. however the two success cases you described didn't work on staging.

**Comment by myriamboure** [comment]
Is it possible to restage so we can have some counter test? Maybe that was a problem with the staging as well, if we haven't the same results in the tests... If some dev can restage we will retest :-)

**Comment by luisramos0** [comment]
staged: https://staging.openfoodfrance.org

**Comment by RachL** [comment]
YES ! Now it is working good : no need for shipping_category in the file, the user guide is correct! Let's roll

**Comment by sigmundpetersen** [comment]
So this should move on to Ready To Go @RachL ?
